hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/InsertAndSelectParser.java

```diff
@@ -0,0 +1,78 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive.persister.entity.impl;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.hibernate.dialect.CockroachDB192Dialect;
+import org.hibernate.dialect.DB2Dialect;
+import org.hibernate.dialect.Dialect;
+import org.hibernate.dialect.PostgreSQL81Dialect;
+import org.hibernate.dialect.SQLServerDialect;
+
+/**
+ * Utility methods to parse SQL and create the string to get the generated id value during an insert.
+ * <p>
+ *     This is an issue with Hibernate ORM 5 and should be unnecessary with Hibernate ORM 6.
+ * <p>
+ */
+// TODO: Add some details about the issue in ORM 5 that prevents us to generate the queries in advance
+public class InsertAndSelectParser {
+
+	private InsertAndSelectParser() {
+	}
+
+	/**
+	 * Matches a SQL server query of this type ({@code default} is optional):
+	 * <pre>
+	 *     insert into entity_table default values select id, name from other_table;
+	 * </pre>
+	 */
+	private static final Pattern MSSQL_PATTERN = Pattern.compile( "^(insert into .*?)(( default)? values.*)" );
+
+	/**
+	 * Queries used to insert a new element and retrieve the id in one go require
+	 * some changes
+	 */
+	public static String parse(String sql, Dialect dialect, String idColumnName) {
+		if ( dialect instanceof PostgreSQL81Dialect || dialect instanceof CockroachDB192Dialect ) {
+			return parsePostgresSQLQuery( sql, idColumnName );
+		}
+		if ( dialect instanceof SQLServerDialect ) {
+			return parseSqlServerQuery( sql, idColumnName );
+		}
+		if ( dialect instanceof DB2Dialect ) {
+			return parseDb2Query( sql, idColumnName );
+		}
+		return sql;
+	}
+
+	/**
+	 * Convert the following SQL query for SQL Server ({@code default} is optional):
+	 * <pre>
+	 *     insert into entity_table default values select id, name from other_table;
+	 * </pre>
+	 * to
+	 * <pre>
+	 *     insert into entity_table output inserted.id default values select id, name from other_table ;
+	 * </pre>
+	 */
+	private static String parseSqlServerQuery(String sql, String idColumnName) {
+		final Matcher matcher = MSSQL_PATTERN.matcher( sql );
+		return matcher.matches()
+				? matcher.group( 1 ) + " output inserted." + idColumnName + matcher.group( 2 )
+				: sql;
+	}
+
+	private static String parseDb2Query(String sql, String idColumnName) {
+		return "select " + idColumnName + " from NEW TABLE (" + sql + ")";
+	}
+
+	private static String parsePostgresSQLQuery(String sql, String idColumnName) {
+		return sql + " returning " + idColumnName;
+	}
```

Comment:
I don't think "parse" is quite the right name for these methods.

Reply:
`convert`? `update`?

Reply:
I will go with `Processor` and `process`, a nice generic term

Reply:
That's fine.